src/core/internal/cppdemangle.d

```diff
@@ -0,0 +1,268 @@
+
+            auto handle = dlopen(null, RTLD_LAZY);
+            assert(handle !is null);
+            auto p = dlsym(handle, "__cxa_demangle");
+
+            if (p !is null)
+            {
+                _handle = handle;
+                _instance.__cxa_demangle = cast(typeof(CPPDemangle.__cxa_demangle)) p;
+                isInizialized = true;
+                return CPPDemangleStatus.INIZIALIZED;
+            }
+
+            dlclose(handle);
+
+            version (OSX)
+                enum names = ["libc++abi.dylib", "libstdc++.dylib"];
+            else version (Posix)
+            {
+                enum names = ["libstdc++.so", "libstdc++.so.6", "libstdc++.so.5",
+                "libc++abi.so", "libc++abi.so.1"];
+            }
+
+            foreach (name; names)
+            {
+                handle = dlopen(name.ptr, RTLD_LAZY);
+
+                if (handle !is null)
+                {
+                    break;
+                }
+            }
+
+            if (handle is null)
+            {
+                return CPPDemangleStatus.LOAD_ERROR;
+            }
+
+            p = dlsym(handle, "__cxa_demangle");
+
+            if (p !is null)
+            {
+                _handle = handle;
+                _instance.__cxa_demangle = cast(typeof(CPPDemangle.__cxa_demangle)) p;
+                isInizialized = true;
+                return CPPDemangleStatus.INIZIALIZED;
+            }
+        }
+
+        version (Windows)
+        {
+            import core.sys.windows.dbghelp : DbgHelp;
+
+            auto dbgHelp = DbgHelp.get();
+
+            if (dbgHelp is null)
+            {
+                return CPPDemangleStatus.LOAD_ERROR;
+            }
+
+            auto func = dbgHelp.UnDecorateSymbolName;
+
+            if (dbgHelp.UnDecorateSymbolName !is null)
+            {
+                _instance.UnDecorateSymbolName = dbgHelp.UnDecorateSymbolName;
+                isInizialized = true;
+                return CPPDemangleStatus.INIZIALIZED;
+            }
+        }
+
+        return CPPDemangleStatus.SYMBOL_ERROR;
+    }
+
+    static CPPDemangle instance() @trusted
+    {
+        return _instance;
+    }
+
+    private char[] _cppdemangle(const(char)[] buf, char[] dst) @trusted
+    {
+        import core.memory : pureCalloc, pureFree;
+        import core.stdc.string : strlen;
+
+        if (!isInizialized)
+        {
+            return null;
+        }
+
+        version (Unicode)
+        {
+            import core.internal.utf : toUTF16z, toUTF8;
+            auto mangledName = toUTF16z(buf);
+        }
+        else
+        {
+            auto mangledName = cast(char*) pureCalloc(buf.length + 1, char.sizeof);
+            scope(exit) pureFree(mangledName);
+            mangledName[0..buf.length] = buf[];
+        }
+
+        version (Posix)
+        {
+            int result;
+            auto demangledNamePtr = _instance.__cxa_demangle(mangledName, null, null, &result);
```

Comment:
I don't see the point of copying the result when you could just pass the pointer to the array here ?